### PR TITLE
add API server

### PR DESCRIPTION
**Pull request by mrd0ll4r**
This is my take on the API server. There are a few open points:

- should the API handle prefixes or should the user handle prefixes?
- do client IDs need to be urlEscaped?
- do we allow a middleware to register handlers for the API?
- should stores return errors for
    - does not exist for deletes?
    - does already exist for inserts?
- should adding/removing an infohash echo back a "sane" version of the infohash, e.g. in hex?

This is WIP, needs discussion.

### CODE REVIEW

**Comment by jzelinskie** [comment]
- I don't think prefixes should be in the store and that the store API should do anymore than present the store interfaces via an HTTP API.
- We should probably add a error in the store for when a resource doesn't exist and update the memory driver to use it.
- Adding handlers to API seems like an interesting idea

File: server/store/memory/ip_store.go [hunk 1]
```diff
@@ -140,8 +140,13 @@ func (s *ipStore) HasAllIPs(ips []net.IP) (bool, error) {
 func (s *ipStore) RemoveIP(ip net.IP) error {
 	s.Lock()
 	defer s.Unlock()
+	key := key(ip)
```
Comment (jzelinskie) [inline server/store/memory/ip_store.go]: let's calculate the key before we lock
woo micro-optimizations

File: server/store/middleware/client/routes.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/chihaya/chihaya/server/store"
+)
+
+// PrefixClient is the prefix to be used for client peer IDs.
+const PrefixClient = "c-"
+
+const pathPutClient = "/client/:client"
+const pathDeleteClient = "/client/:client"
+
+var routesActivated = &sync.Once{}
```
Comment (jzelinskie) [inline server/store/middleware/client/routes.go]: Is there a reason why this isn't just `var routesActivated sync.Once`?

Comment (mrd0ll4r) [inline server/store/middleware/client/routes.go]: Just me programming way too much C(++) lately, will fix...

**Comment by mrd0ll4r** [comment]
Should the store API provide hasIP/hasNetwork, getSeeders/getLeechers, numSeeders/numLeechers,... as well?

**Comment by jzelinskie** [comment]
I think it should.

This PR is looking pretty good. It's got a lot of good stuff going on, it just appears to be 85% complete.
If you used `TODO`s into your code, I'd be more willing to merge things that are still in progress.

Some tiny nitpicks: 
- REST resources should be plural i.e. /network**s**/:network, ip**s**/:ip.
- code blocks should use triple upticks instead of indentation
- Let's rename ErrDNE to `ErrResourceDoesNotExist`

**Comment by mrd0ll4r** [comment]
I added a few of those, namely matching IPs, infohashes and clientIDs. I'm not sure about the endpoints for all the peer store methods, so I'll leave that out for now.

**Comment by jzelinskie** [comment]
You need to make a new PR with this branch pointed at master

**Pull request closed without merging**